# Exploring eBay Car Sales Data

## Introduction

In this project we are gonna clean the data and analyze the included car listing.

The project instruction gives the column of the dataset. Let's look at the columns of the data set.

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* odometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

We are starting with the importing the libraries.

In [1]:
import pandas as pd

autos = pd.read_csv("autos.csv", encoding ="Latin")

In [2]:
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

We call our datase as you can see. So, rigth now we need to know general info for this dataset. 

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

We find out only 5 "int64" data type in this dataset and and rest of the columns which is 15 is "object" data type. And also we have 20 columns for this dataset.

After that, we will see the first 5 rows of the dataset.

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

After all this step, we should see the columns clearly with ".columns" method.

## Cleaning Column Names

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
# Changing the a few columns names

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']


# But if the instructions didn't say the "the rest of the column names from camelcase to snakecase", we can use next code piece for the changing columns names.

# autos.rename({"yearOfRegistration" : "registration_year"}, axis = 1,  inplace = True)

# autos.rename({"monthOfRegistration" : "registration_month"}, axis = 1,  inplace = True)

# autos.rename({"notRepairedDamage" : "unrepaired_damage"}, axis = 1,  inplace = True)

# autos.rename({"dateCreated" : "ad_created"}, axis = 1,  inplace = True)


Let's check the changed columns names with two methods.

In [8]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Initial Exploration and Cleaning

Now it's time to know descriptive statics for all columns.

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

So, we need to see also only numeric columns.

In [11]:
autos.describe()

registration_year      power_ps  registration_month  num_photos  \
count       50000.000000  50000.000000        50000.000000     50000.0   
mean         2005.073280    116.355920            5.723360         0.0   
std           105.712813    209.216627            3.711984         0.0   
min          1000.000000      0.000000            0.000000         0.0   
25%          1999.000000     70.000000            3.000000         0.0   
50%          2003.000000    105.000000            6.000000         0.0   
75%          2008.000000    150.000000            9.000000         0.0   
max          9999.000000  17700.000000           12.000000         0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

We should answer for three things. These are:
1. Any columns that have mostly one value that are candidates to be dropped
2. Any columns that need more investigation.
3. Any examples of numeric data stored as text that needs to be cleaned.

**1- Any columns that have mostly one value that are candidates to be dropped.**

- ***num_photos column*** : All values of this column appear to be zero. If this is true, this column does not contain much information and can be removed from the data set.

**2- Any columns that need more investigation.** 

- ***registration_year column***: This column appears to have values that are too large or too small (for example, 9999 or 1000). It may be necessary to examine this situation in more detail.

- ***registration_month column***: 0 month value stands out. By examining this value, it is important to identify missing or incorrect data.

**3- Any examples of numeric data stored as text that needs to be cleaned.**

- ***power_ps column***: Some values appear to be stored as text. It may be necessary to convert this data into numerical format.

So, let's examine the "num_photos" columns further more.

In [12]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

It looks like the num_photos column has 0 for every column that's why we will drop this column.

In [13]:
autos = autos.drop(["num_photos"], axis = 1)

And also there are two columns which are numeric values bu stored as text. So we need the convert these two columns to numeric datatype.
We know any of the data in "price" column is starting with "$" and contains ",". We need to remove them from the datas. 

In [14]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

Now we are gonna convert the "odometer" column to "odometer_km". And also we know any of the data in "odometer" column is starting with "km" and contains ",". We need to remove them from the datas. 

In [15]:
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Exploring the Odometer and Price Column

We already change the object data type to numeric data type for these columns. 

In [16]:
autos["price"].unique().shape

(2357,)

We get the number of ht eunique values of the "price" column. There is 2357 unique data.

Now let's get the descriptive analysis for this column.

In [17]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

After that, we can see minimum data for this column is 0 and rest of the calculation.

In [18]:
autos["price"].value_counts().head(10)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64

And let's see the value counts descending order.

From here we understand that there are 1421 cars with a price of 0. We know that there are 50000 pieces of data in this data set. We can deduce that this value we obtained is approximately 2% of the entire data set. 

In [19]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

As can be seen, when we look at the highest prices, there is a sudden and high price increase from 350000 to 999990. While values up to 350000 show a gradual increase, after this value there is a sudden jump to less realistic figures. Therefore, it would be a very correct decision to eliminate values above 350000.

In [20]:
autos = autos[autos["price"].between(1,35010)]
autos["price"].describe()

count    48056.000000
mean      5331.427168
std       6047.229364
min          1.000000
25%       1200.000000
50%       2999.000000
75%       7200.000000
max      35000.000000
Name: price, dtype: float64

## Exploring The Date Columns

So let's examine the date column.

- `date_crawled`
- `last_seen`
- `ad_created`
- `registration_month`
- `registration_year`

In [21]:
autos[["date_crawled", "last_seen", "ad_created", "registration_month", "registration_year"]][0:5]

date_crawled            last_seen           ad_created  \
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00   
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00   
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00   
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00   
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

We get the exalty which columns is in date format. So there is 3 date format columns. These are "date_crawled", "ad_created" and "last_seen". 

In [22]:
date_crawled_freq = (autos["date_crawled"]
        .str[:10] # like this 2016-03-05 
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )
date_crawled_freq

2016-03-05    0.025366
2016-03-06    0.014067
2016-03-07    0.036083
2016-03-08    0.033378
2016-03-09    0.033149
2016-03-10    0.032233
2016-03-11    0.032483
2016-03-12    0.037082
2016-03-13    0.015690
2016-03-14    0.036603
2016-03-15    0.034293
2016-03-16    0.029653
2016-03-17    0.031630
2016-03-18    0.012860
2016-03-19    0.034689
2016-03-20    0.037872
2016-03-21    0.037144
2016-03-22    0.032920
2016-03-23    0.032233
2016-03-24    0.029278
2016-03-25    0.031526
2016-03-26    0.032379
2016-03-27    0.031047
2016-03-28    0.034793
2016-03-29    0.034189
2016-03-30    0.033752
2016-03-31    0.031775
2016-04-01    0.033607
2016-04-02    0.035438
2016-04-03    0.038684
2016-04-04    0.036582
2016-04-05    0.012922
2016-04-06    0.003184
2016-04-07    0.001415
Name: date_crawled, dtype: float64

When we looked at these data, we reached the frequencies of the dates in the data set. When we look at our output, we see that the dates are in order but the frequency values are not in order or orderly, we can make clearer and better inferences if we sort them.

In [23]:
date_crawled_freq.sort_values()

2016-04-07    0.001415
2016-04-06    0.003184
2016-03-18    0.012860
2016-04-05    0.012922
2016-03-06    0.014067
2016-03-13    0.015690
2016-03-05    0.025366
2016-03-24    0.029278
2016-03-16    0.029653
2016-03-27    0.031047
2016-03-25    0.031526
2016-03-17    0.031630
2016-03-31    0.031775
2016-03-23    0.032233
2016-03-10    0.032233
2016-03-26    0.032379
2016-03-11    0.032483
2016-03-22    0.032920
2016-03-09    0.033149
2016-03-08    0.033378
2016-04-01    0.033607
2016-03-30    0.033752
2016-03-29    0.034189
2016-03-15    0.034293
2016-03-19    0.034689
2016-03-28    0.034793
2016-04-02    0.035438
2016-03-07    0.036083
2016-04-04    0.036582
2016-03-14    0.036603
2016-03-12    0.037082
2016-03-21    0.037144
2016-03-20    0.037872
2016-04-03    0.038684
Name: date_crawled, dtype: float64

Now we get the sorted values. We can determine that **"2016-04-03"** is the date with the highest number of logins. This date accounts for *3.87%* of the data set. Additionally, it would be a correct determination to determine **"2016-04-07"** as the day with the lowest number of entries. On this day, the density value was obtained as *0.14%*. 

Right now we are gonn apply same method for "last_seen" column.

In [24]:
(autos["last_seen"]
        .str[:10] # like this 2016-03-05 
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )


2016-03-05    0.001082
2016-03-06    0.004370
2016-03-07    0.005390
2016-03-08    0.007491
2016-03-09    0.009655
2016-03-10    0.010758
2016-03-11    0.012465
2016-03-12    0.023930
2016-03-13    0.008948
2016-03-14    0.012631
2016-03-15    0.015981
2016-03-16    0.016460
2016-03-17    0.028196
2016-03-18    0.007346
2016-03-19    0.015919
2016-03-20    0.020643
2016-03-21    0.020663
2016-03-22    0.021475
2016-03-23    0.018603
2016-03-24    0.019810
2016-03-25    0.019269
2016-03-26    0.016814
2016-03-27    0.015690
2016-03-28    0.020996
2016-03-29    0.022474
2016-03-30    0.024908
2016-03-31    0.023826
2016-04-01    0.022911
2016-04-02    0.025012
2016-04-03    0.025283
2016-04-04    0.024659
2016-04-05    0.124001
2016-04-06    0.221346
2016-04-07    0.130993
Name: last_seen, dtype: float64

As can be seen, there are quite high differences between the last three data and the remaining data. It is clearly seen that there is a disproportionate ratio. The value of the date "2016-04-04" was determined to be 2.46%, but the value of the following date "2016-04-06" was found to be 22.13%. The difference determined between these two values is almost 10 times. It is unusual for these ads to have been removed at such a high rate.

And there is one left date column.

In [25]:
(autos["ad_created"]
        .str[:10] # like this 2016-03-05 
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
                ...   
2016-04-03    0.038913
2016-04-04    0.036978
2016-04-05    0.011632
2016-04-06    0.003267
2016-04-07    0.001269
Name: ad_created, Length: 75, dtype: float64

In [26]:
autos["registration_year"].describe()

count    48056.000000
mean      2004.635862
std         87.023293
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We see that the minimum value is 1000, long before automobiles were invented, and the maximum value is 9999, many years later. These values are unacceptable as history.

## Dealing With Incorrect Registration Year Data

As we examined, the smallest value in the dates was 1000 and the largest value was 9999, but these values were considered invalid for this data set. When examining the data set, we need to find a valid date range. 

Any vehicle with a registration year over 2016 is at fault, as a car cannot be registered for the first time after the advertisement is seen, but finding the earliest date is a bit problematic.

To stay on the safe side, let's check how many vehicles have registration years that fall outside the range of 1900 to 2016. Afterwards, we can decide if this represents a significant number.

In [27]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.03905859830198102

What we do in this code cell is to find out the percentage of the data set for years that are not between 1900 and 2016. We see that we get a rate close to 4%.

In [28]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).sort_index().tail(25)

1992    0.007947
1993    0.009138
1994    0.013513
1995    0.026462
1996    0.029689
1997    0.042162
1998    0.051084
1999    0.062712
2000    0.068299
2001    0.057039
2002    0.053747
2003    0.058425
2004    0.058338
2005    0.063492
2006    0.057732
2007    0.048918
2008    0.047489
2009    0.044912
2010    0.033825
2011    0.034345
2012    0.026787
2013    0.016111
2014    0.012928
2015    0.006778
2016    0.026116
Name: registration_year, dtype: float64

The most common records were observed after 2000. We can say that the most intense period is 1997 and 2009. At the same time, we can say that the frequencies were low after 1997. And a sudden and high decline is observed in 2015.

## Exploring Price By Brand

In [29]:
brands_count = autos["brand"].value_counts(normalize=True)
brands_count

volkswagen        0.212867
bmw               0.109769
opel              0.108729
mercedes_benz     0.095498
audi              0.085299
ford              0.070443
renault           0.047641
peugeot           0.030165
fiat              0.025921
seat              0.018472
skoda             0.016588
nissan            0.015440
mazda             0.015332
smart             0.014314
citroen           0.014162
toyota            0.012820
hyundai           0.010134
volvo             0.009247
sonstige_autos    0.009247
mini              0.008857
mitsubishi        0.008294
honda             0.007904
kia               0.007146
alfa_romeo        0.006691
suzuki            0.005998
chevrolet         0.005717
chrysler          0.003551
porsche           0.003378
dacia             0.002664
daihatsu          0.002534
subaru            0.002165
jeep              0.002144
land_rover        0.001754
saab              0.001667
daewoo            0.001516
jaguar            0.001516
trabant           0.001408
r

Here we see that there are 40 different car brands in this data set. As can be seen, the top 5 brands with the highest density are based in Germany. We can deduce that the frequency density of these brands is between 50% and 60% in the data set in total.

As we mentioned, there are 40 different car brands in this data set. These brands have very small concentrations. We can ignore these small intensities.

In [30]:
high_brands = brands_count[brands_count > .05].index

high_brands


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [32]:
mean_of_brand = {}
;
for brand in high_brands:
    brand_name = autos[autos["brand"] == brand]
    mean_brand = brand_name["price"].mean()
    mean_of_brand[brand] = int(mean_brand)
    
mean_of_brand



{'volkswagen': 5280,
 'bmw': 7742,
 'opel': 2968,
 'mercedes_benz': 7727,
 'audi': 8377,
 'ford': 3577}

If we look at the averages we obtained, our inferences can be listed as follows.
1. The most expensive brand is `Audi`,
2. `BMW` and `Mercedes Benz` can also be among the expensive brands,
3. Low brands consist of `Opel` and `Ford` brands,
4. `Volkswagen` brand is neither expensive nor cheap. We can say that this brand is the preferred option between these two parties.

## Storing Aggregate Data in DataFrame

In [33]:
high_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [34]:
mileage_mean = {}

for mile in high_brands:
    brand_name = autos[autos["brand"] == mile]
    mean_brand = brand_name["odometer_km"].mean()
    mileage_mean[mile] = int(mean_brand)
    
mileage_mean


{'volkswagen': 128961,
 'bmw': 133689,
 'opel': 129310,
 'mercedes_benz': 132235,
 'audi': 131315,
 'ford': 124495}

In [35]:
mean_mileage = pd.Series(mileage_mean).sort_values(ascending=False)
mean_prices = pd.Series(mean_of_brand).sort_values(ascending=False)

In [36]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  133689
mercedes_benz        132235
audi                 131315
opel                 129310
volkswagen           128961
ford                 124495

In [37]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  133689        7742
mercedes_benz        132235        7727
audi                 131315        8377
opel                 129310        2968
volkswagen           128961        5280
ford                 124495        3577